cpp/src/arrow/engine/substrait/serde.h

```diff
@@ -40,22 +41,81 @@ using ConsumerFactory = std::function<std::shared_ptr<compute::SinkNodeConsumer>
+/// \param[out] ext_set_out if non-null, the extension mapping used by the Substrait
+/// Plan is returned here.
+/// \return an ExecNode corresponding to the single toplevel relation in the Substrait
+/// Plan
 Result<compute::ExecPlan> DeserializePlan(const Buffer& buf,
                                           const ConsumerFactory& consumer_factory,
+                                          const ExtensionIdRegistry* registry = NULLPTR,
                                           ExtensionSet* ext_set_out = NULLPTR);
 
+/// Factory function type for generating the write options of a node consuming the batches
+/// produced by each toplevel Substrait relation when deserializing a Substrait Plan.
+using WriteOptionsFactory = std::function<std::shared_ptr<dataset::WriteNodeOptions>()>;
+
+/// \brief Deserializes a Substrait Plan message to a list of ExecNode declarations
+///
+/// The output of each top-level Substrait relation will be written to a filesystem.
+/// `write_options_factory` can be used to control write behavior.
+///
+/// \param[in] buf a buffer containing the protobuf serialization of a Substrait Plan
+/// message
+/// \param[in] write_options_factory factory function for generating the write options of
+/// a node consuming the batches produced by each toplevel Substrait relation
+/// \param[in] registry an extension-id-registry to use, or null for the default one.
+/// \param[out] ext_set_out if non-null, the extension mapping used by the Substrait
+/// Plan is returned here.
+/// \return a vector of ExecNode declarations, one for each toplevel relation in the
+/// Substrait Plan
+ARROW_ENGINE_EXPORT Result<std::vector<compute::Declaration>> DeserializePlans(
+    const Buffer& buf, const WriteOptionsFactory& write_options_factory,
+    const ExtensionIdRegistry* registry = NULLPTR, ExtensionSet* ext_set_out = NULLPTR);
+
+/// \brief Deserializes a single-relation Substrait Plan message to an execution plan
+///
+/// The output of the single Substrait relation will be written to a filesystem.
+/// `write_options_factory` can be used to control write behavior.
+///
+/// \param[in] buf a buffer containing the protobuf serialization of a Substrait Plan
+/// message
+/// \param[in] write_options_factory factory function for generating the write options of
+/// a node consuming the batches produced by each toplevel Substrait relation
+/// \param[in] registry an extension-id-registry to use, or null for the default one.
+/// \param[out] ext_set_out if non-null, the extension mapping used by the Substrait
+/// Plan is returned here.
+/// \return a vector of ExecNode declarations, one for each toplevel relation in the
+/// Substrait Plan
+ARROW_ENGINE_EXPORT Result<compute::ExecPlan> DeserializePlan(
+    const Buffer& buf, const WriteOptionsFactory& write_options_factory,
+    const ExtensionIdRegistry* registry = NULLPTR, ExtensionSet* ext_set_out = NULLPTR);
```

Comment:
For both this and the other `DeserializePlan` method it doesn't make too much sense for the user to have to provide an entire factory.

Instead could we change `const WriteOptionsFactory& write_options_factory` to `dataset::WriteNodeOptions write_options`?

We should also change the other `DeserializePlan` to use `std::shared_ptr<compute::SinkNodeConsumer>` instead of `const ConsumerFactory&` but that could be part of a future cleanup.

Reply:
I think changing from factories to instances would lead to trouble, `ConsumerFactory` (and `WriteOptionsFactory` too) is called once per `PlanRel` in the Substrait plan in order to get a distinct instance per invocation, despite the factory taking no arguments, which often leads to returning equal instances. However, using an instance instead of a factory would pretty much force reusing the same instance with each `PlanRel`, which isn't right. To avoid reusing such an instance, it would have to support safe duplication, which would be more annoying for the caller to implement than a factory.

Reply:
This would only be for the `DeserializePlan` methods which should fail if there is more than one `PlanRel` anyways so I'm not sure it is an issue?  No copies would need to be made of the options.

Reply:
Ah, I didn't notice it was the singular variation. The code currently has the singular variation invoke the plural one, such that the change you're asking is somewhat inconvenient to make, but I'll try fixing for the sake of user convenience.

Reply:
Done.